File: testbed/test/cruise/associations/OptionalOneToOneTest.java
package cruise.associations;

import org.junit.*;

import cruise.associations.MentorB;
import cruise.associations.ProgramB;
import cruise.associations.StudentB;

public class OptionalOneToOneTest
{

  @Test
  public void Constructor()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    
    Assert.assertEquals(student,mentor.getStudent());
    Assert.assertEquals(mentor,student.getMentor());
  } 
  
  @Test
  public void CannotSetStudentBToNull()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    
    Assert.assertEquals(false,mentor.setStudent(null));
  }
    
  @Test
  public void SetStudent()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    
    StudentB student2 = new StudentB();
    Assert.assertEquals(true, mentor.setStudent(student2));
    
    
    Assert.assertEquals(mentor,student2.getMentor());
    Assert.assertEquals(student2,mentor.getStudent());
    Assert.assertEquals(null,student.getMentor());
  } 
  
  @Test
  public void SetMentorCannotReset()
  {
    StudentB s = new StudentB();
    MentorB m = new MentorB(s);
    
    StudentB s2 = new StudentB();
    MentorB m2 = new MentorB(s2);
    
    Assert.assertEquals(false,m.setStudent(s2));
    Assert.assertEquals(m,s.getMentor());
    Assert.assertEquals(s,m.getStudent());
    Assert.assertEquals(m2,s2.getMentor());
    Assert.assertEquals(s2,m2.getStudent());
  } 
  
  @Test
  public void SetMentor()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    
    StudentB student2 = new StudentB();
    student2.setMentor(mentor);
    
    Assert.assertEquals(mentor,student2.getMentor());
    Assert.assertEquals(student2,mentor.getStudent());
    Assert.assertEquals(null,student.getMentor());
  }   
  
  @Test(expected=RuntimeException.class)
  public void UnableToConstructNewSubordinateFromExistingDriverThatAlreadyHasDriver()
  {
    StudentB student = new StudentB();
    new MentorB(student);
    new MentorB(student); 
  } 

  @Test
  public void DeleteDriverStudentHasNoMentor()
  {
    StudentB student = new StudentB();
    student.delete();
    Assert.assertEquals(null, student.getMentor());
  }
  
  @Test
  public void DeleteDriverRemovesSubordinate()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    ProgramB program = new ProgramB();
    mentor.setProgram(program);
    
    student.delete();
    Assert.assertEquals(null, student.getMentor());
    Assert.assertEquals(null, mentor.getStudent());
    Assert.assertEquals(null, mentor.getProgram());
  }  
  
  @Test
  public void DeleteSubordinateKeepDriver()
  {
    StudentB student = new StudentB();
    MentorB mentor = new MentorB(student);
    ProgramB program = new ProgramB();
    student.setProgram(program);
    
    mentor.delete();
    Assert.assertEquals(null, student.getMentor());
    Assert.assertEquals(null, mentor.getStudent());
    Assert.assertEquals(program, student.getProgram());
  }  
  
}

File: testbed_python/test/associations/optional_one_to_one_test.py
import unittest

# Test classes are imported dynamically
# Must be generated into ImportModules' namespace then imported
from ImportModules import importModules

importModules(["MentorB", "ProgramB", "StudentB"], ["associations"])
from ImportModules import *


class OptionalOneToOneTest(unittest.TestCase):
    def test_constructor(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)

        self.assertEqual(student, mentor.getStudent())
        self.assertEqual(mentor, student.getMentor())

    def test_cannotSetStudentBToNull(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)

        self.assertEqual(False, mentor.setStudent(None))

    @unittest.skip(".equals()")
    def test_setStudent(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)

        student2 = StudentB.StudentB()
        self.assertEqual(True, mentor.setStudent(student2))

        self.assertEqual(mentor, student2.getMentor())
        self.assertEqual(student2, mentor.getStudent())
        self.assertEqual(None, student.getMentor())

    def test_setMentorCannotReset(self):
        s = StudentB.StudentB()
        m = MentorB.MentorB(s)

        s2 = StudentB.StudentB()
        m2 = MentorB.MentorB(s2)

        self.assertEqual(False, m.setStudent(s2))
        self.assertEqual(m, s.getMentor())
        self.assertEqual(s, m.getStudent())
        self.assertEqual(m2, s2.getMentor())
        self.assertEqual(s2, m2.getStudent())

    @unittest.skip(".equals()")
    def test_setMentor(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)

        student2 = StudentB.StudentB()
        student2.setMentor(mentor)

        self.assertEqual(mentor, student2.getMentor())
        self.assertEqual(student2, mentor.getStudent())
        self.assertEqual(None, student.getMentor())

    def test_unableToConstructNewSubordinateFromExistingDriverThatAlreadyHasDriver(
        self,
    ):
        with self.assertRaises(RuntimeError):
            student = StudentB.StudentB()
            MentorB.MentorB(student)
            MentorB.MentorB(student)

    def test_deleteDriverRemovesSubordinate(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)
        program = ProgramB.ProgramB()
        mentor.setProgram(program)

        del student
        self.assertFalse(mentor.getStudent() in locals())
        self.assertFalse(mentor.getProgram() in locals())

    def test_deleteSubordinateKeepDriver(self):
        student = StudentB.StudentB()
        mentor = MentorB.MentorB(student)
        program = ProgramB.ProgramB()
        student.setProgram(program)

        del mentor
        self.assertFalse(student.getMentor() in locals())
        self.assertEqual(program, student.getProgram())
